microsoft/planetary-computer-apis
# Enable auto-scaling

## Description

This PR contains the code and configuration used in an auto-scaling environment. It has the objective of scale the cluster based on the amount of HTTP requests received by any API developed with FastAPI, in this scenario, the Stac API.

- Auto-scaling environment deployment.
- Enabled HTTP Total Requests from Stac API using Prometheus.
- Prometheus Adapter enabled.
- HPA enabled reading http_total_request metric from Stac API.
- Overprovisioning strategy.

The idea from this PR is to test and ensure 

## Type of change

Please delete options that are not relevant.


## How Has This Been Tested?

The environment was deployed and tested using Azure Load Testing.

Load Test
![bench_v4_hpa_overprov-test](https://github.com/microsoft/planetary-computer-apis/assets/26978598/a0151e5d-cee4-4905-bd49-8a1bdbd3dfd3)

In green http requests received by the Stac API. Blue lines means the start and end of the load test.
![bench_v4_hpa_overprov-http](https://github.com/microsoft/planetary-computer-apis/assets/26978598/0d27d0a7-6140-48c9-bdeb-9f0f31e5d47c)

Replicas and Nodes during test.
![bench_v4_hpa_overprov-pods](https://github.com/microsoft/planetary-computer-apis/assets/26978598/b6c953a5-7570-4afc-a7ab-cb657104a03e)


How to test:
1. Deploy the environment.
2. Run any of the Benchmarks on the Azure Load Testing instance.
3. Navigate to the Cluster instance on Azure Portal and check that there should be several pods of the Stac API running. 

Another option is to activate Managed Grafana and visualize metrics there.

## Checklist:

Please delete options that are not relevant.
